### PR TITLE
:wrench: Read component shapes from pages

### PR DESCRIPTION
**Pull request by hirunatan**
WARNING: big refactor, may generate some unwanted effects. We need to talk before merging.

https://tree.taiga.io/project/penpot/task/4256

There is a TODO in `rpc/commands/files.clj` about [loading libraries](https://tree.taiga.io/project/penpot/task/4904) with their full data, that needs to be rethought, but may be done after merging this.

### CODE REVIEW

File: backend/src/app/rpc/commands/files.clj [hunk 1]
```diff
@@ -552,7 +555,10 @@
        (map (fn [{:keys [id] :as row}]
               (binding [pmap/*load-fn* (partial load-pointer conn id)]
                 (-> row
-                    (update :data dissoc :pages-index)
+                    ;; TODO: re-enable this dissoc and replace call
```
Comment (hirunatan) [inline backend/src/app/rpc/commands/files.clj:572]: May ignore this, i'm working on it in other branch

File: common/src/app/common/pages/helpers.cljc [hunk 2]
```diff
@@ -249,25 +249,6 @@
   [shape group]
   ((or (:touched shape) #{}) group))
 
-(defn get-component
```
Comment (hirunatan) [inline common/src/app/common/pages/helpers.cljc:273]: I'm gradually moving helpers from this file to common.types

File: common/src/app/common/types/file.cljc [hunk 3]
```diff
@@ -116,33 +116,105 @@
   ([libraries library-id component-id]
    (ctkl/get-component (dm/get-in libraries [library-id :data]) component-id)))
 
+(defn get-component-library
```
Comment (hirunatan) [inline common/src/app/common/types/file.cljc:119]: Added a bunch of helpers here to manage components reading their content from the corresponding page of a file (or the component itself if still in v1)

File: common/src/app/common/types/file.cljc [hunk 4]
```diff
@@ -477,69 +559,80 @@
...
-                                 (let [component-id      (:component-id shape)
-                                       component-file-id (:component-file shape)
-                                       component-file    (when component-file-id (get libraries component-file-id nil))
-                                       component         (if component-file
-                                                           (dm/get-in component-file [:data :components component-id])
-                                                           (get components component-id))]
-                                   (str/format " (%s%s)"
-                                               (when component-file (str/format "<%s> " (:name component-file)))
-                                               (:name component))))))))]
+                         (let [shape (get objects shape-id)]
```
Comment (hirunatan) [inline common/src/app/common/types/file.cljc:534]: Deep review of this function to display the new components structure

File: frontend/src/app/main/data/workspace.cljs [hunk 5]
```diff
@@ -233,6 +233,19 @@
                                   (->> (filter (comp t/pointer? val) data)
                                        (resolve-pointers id)
                                        (rx/map #(update file :data merge %)))))
+                               (rx/mapcat
```
Comment (hirunatan) [inline frontend/src/app/main/data/workspace.cljs:236]: Now when loading libraries need to load all the pages, and resolve the fragments too (previously only assets were loaded)

File: common/src/app/common/types/file.cljc [hunk 6]
```diff
@@ -116,33 +116,105 @@
...
-                                          :main-instance-x (:x main-instance)   ; An instance root is always a group,
-                                          :main-instance-y (:y main-instance))] ; so it will have :x and :y
-                 (when (nil? main-instance)
-                   (throw (ex-info "Cannot delete the main instance before the component" {:component-id component-id})))
-                 (assoc-in file-data [:deleted-components component-id] component))
-               file-data)))]
+  ([file-data component-id _skip-undelete?]
+   (let [_components-v2 (dm/get-in file-data [:options :components-v2])
+
+        ;; TODO: replace :deleted-components with a :deleted? flag in normal shapes
```
Comment (hirunatan) [inline common/src/app/common/types/file.cljc]: Done in a later commit

File: backend/src/app/rpc/commands/files.clj [hunk 7]
```diff
@@ -492,10 +492,13 @@
 
           (library-summary [{:keys [id data] :as file}]
             (binding [pmap/*load-fn* (partial load-pointer conn id)]
-              {:components (assets-sample (:components data) 4)
-               :media (assets-sample (:media data) 3)
-               :colors (assets-sample (:colors data) 3)
-               :typographies (assets-sample (:typographies data) 3)}))]
+              (let [components-sample (-> (assets-sample (:components data) 4)
+                                          (update :sample
+                                                  #(map (partial ctf/load-component-objects data) %)))]
```
Comment (pabloalba) [inline backend/src/app/rpc/commands/files.clj]: Is this safe with components-v1 ?

File: common/src/app/common/types/container.cljc [hunk 8]
```diff
@@ -62,6 +63,17 @@
   [container shape-id f]
   (update-in container [:objects shape-id] f))
 
+(defn get-component-shape
```
Comment (pabloalba) [inline common/src/app/common/types/container.cljc:66]: In file.cljs there is a helper with the same name that does another thing... Maybe it's better to use another name?

File: common/src/app/common/data.cljc [hunk 9]
```diff
@@ -477,6 +482,17 @@
   (->> (apply c/iteration args)
        (concat-all)))
 
+(defn insert-at-index
+  "Insert a list of elements at the given index of a previous list.
+  Replace all existing elems."
+  [elems index new-elems]
+  (let [[before after] (split-at index elems)
+        p? (set new-elems)]
```
Comment (pabloalba) [inline common/src/app/common/data.cljc:490]: why `p?` ? It is a set, not a boolean

File: common/src/app/common/types/file.cljc [hunk 10]
```diff
@@ -171,73 +162,54 @@
...
+  "Recover a deleted component and all its shapes and put all this again in place."
   [file-data component-id]
-  (let [component (-> (dm/get-in file-data [:deleted-components component-id])
-                      (dissoc :main-instance-x :main-instance-y))]
-    (cond-> file-data
-      (some? component)
-      (-> (assoc-in [:components component-id] component)
-          (d/dissoc-in [:deleted-components component-id])))))
+  (-> file-data
+      (ctkl/update-component component-id #(dissoc % :objects))
```
Comment (pabloalba) [inline common/src/app/common/types/file.cljc:206]: Doesn't this break v1 componentes?

File: common/src/app/common/data.cljc (hunk 9 above)
Comment (hirunatan) [reply common/src/app/common/data.cljc:490]: this is the clojure idiom when we call a set as if it was a function, that returns true if the argument is inside the set

File: common/src/app/common/types/container.cljc (hunk 8 above)
Comment (hirunatan) [reply common/src/app/common/types/container.cljc:66]: It's possible. When we drop the components-v1 compatibility and remove container abstraction, I want to do a general review of helpers. Until then, I'm documenting all this and thinking of a glossary of terms, and will review many names.

File: common/src/app/common/types/file.cljc (hunk 10 above)
Comment (hirunatan) [reply common/src/app/common/types/file.cljc:206]: In v1 there is no restore component. This feature only exists in v2.

File: backend/src/app/rpc/commands/files.clj (hunk 7 above)
Comment (hirunatan) [reply backend/src/app/rpc/commands/files.clj]: Yes, I think. load-component-objects checks for v1, and if so, returns the component untouched. Is there any other problem you see?

Comment (pabloalba) [reply backend/src/app/rpc/commands/files.clj]: I was concern because there were no check for v2, but if it's done in load-component-objects, then is ok

**Pull request merged**